cfg/config.go

```diff
@@ -0,0 +1,184 @@
+/*
+ * knoxite
+ *     Copyright (c) 2020, Nicolas Martin <penguwin@penguwin.eu>
+ *
+ *   For license see LICENSE
+ */
+package cfg
+
+import (
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	gap "github.com/muesli/go-app-paths"
+)
+
+const appName = "knoxite"
+
+var cfgFileName = "knoxite.conf"
+
+// The RepoConfig struct contains all the default values for a a repository
+type RepoConfig struct {
+	Url         string `json:"url"`
+	Compression string `json:"compression"`
+	Tolerance   uint   `json:"tolerance"`
+	Encryption  string `json:"encryption"`
+}
+
+type Config struct {
+	Repositories map[string]RepoConfig
+	backend      ConfigBackend
+	url          *url.URL
+}
+
+// ConfigBackend is the interface implemented by the configuration backends.
+//
+// Backends are responsible for loading and saving the Config struct to
+// the local filesystem, the network, etc.
+type ConfigBackend interface {
+	Load(*url.URL) (*Config, error)
+	Save(*Config) error
+}
+
+func (c *Config) Save() error {
+	return c.backend.Save(c)
+}
+
+// Load the configuration.
+//
+// The backend loaded will be responsible for loading it
+// from the given URL.
+func (c *Config) Load() error {
+	config, err := c.backend.Load(c.url)
+	if err != nil {
+		return err
+	}
+	c.Repositories = config.Repositories
+	return nil
+}
+
+// Backend currently being used.
+func (c *Config) Backend() ConfigBackend {
+	return c.backend
```

Comment:
Any reason you use an getter for c.backend instead of making the backend public? 
You could also use `json:"-"?

Reply:
I think we probably shouldn't expose the backend at all. There's no good reason I see for anyone to call it directly. The only callers of this method are currently the config tests. These tests are valid, but we can also achieve them in a different way:

Add a `Type() int` method to the `ConfigBackend` interface, so we can compare which type of `ConfigBackend` we're dealing with.

Reply:
The `ConfigBackend` interface now provides a `Type() int` method.
Currenlty we have these configration backends available:
```
const (
	CryptoConf = iota
	MemoryConf
	FileConf
)
```